#Домашнє завдання до тем apply(), groupby(), pivot_table()

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [3]:
import pandas as pd
data_path = '../Data/supermarket_sales.csv'
df = pd.read_csv(data_path)

In [4]:
df.head(10)

Invoice ID Branch       City Customer type  Gender  \
0  750-67-8428      A     Yangon        Member  Female   
1  226-31-3081      C  Naypyitaw        Normal  Female   
2  631-41-3108      A     Yangon        Normal    Male   
3  123-19-1176      A     Yangon        Member    Male   
4  373-73-7910      A     Yangon        Normal    Male   
5  699-14-3026      C  Naypyitaw        Normal    Male   
6  355-53-5943      A     Yangon        Member  Female   
7  315-22-5665      C  Naypyitaw        Normal  Female   
8  665-32-9167      A     Yangon        Member  Female   
9  692-92-5582      B   Mandalay        Member  Female   

             Product line  Unit price  Quantity   Tax 5%     Total       Date  \
0       Health and beauty       74.69         7  26.1415  548.9715   1/5/2019   
1  Electronic accessories       15.28         5   3.8200   80.2200   3/8/2019   
2      Home and lifestyle       46.33         7  16.2155  340.5255   3/3/2019   
3       Health and beauty       58.22         8  23.2880  489.0480  1/27/2019   
4       Sports and travel       86.31         7  30.2085  634.3785   2/8/2019   
5  Electronic accessories       85.39         7  29.8865  627.6165  3/25/2019   
6  Electronic accessories       68.84         6  20.6520  433.6920  2/25/2019   
7      Home and lifestyle       73.56        10  36.7800  772.3800  2/24/2019   
8       Health and beauty       36.26         2   3.6260   76.1460  1/10/2019   
9      Food and beverages       54.84         3   8.2260  172.7460  2/20/2019   

    Time      Payment    cogs  gross margin percentage  gross income  Rating  
0  13:08      Ewallet  522.83                 4.761905       26.1415     9.1  
1  10:29         Cash   76.40                 4.761905        3.8200     9.6  
2  13:23  Credit card  324.31                 4.761905       16.2155     7.4  
3  20:33      Ewallet  465.76                 4.761905       23.2880     8.4  
4  10:37      Ewallet  604.17                 4.761905       30.2085     5.3  
5  18:30      Ewallet  597.73                 4.761905       29.8865     4.1  
6  14:36      Ewallet  413.04                 4.761905       20.6520     5.8  
7  11:38      Ewallet  735.60                 4.761905       36.7800     8.0  
8  17:15  Credit card   72.52                 4.761905        3.6260     7.2  
9  13:27  Credit card  164.52                 4.761905        8.2260     5.9

1. Дослідимо, який філіал супермаркету ('Branch') є найприбутковішим. Для цього знайдіть сумарний прибуток за кожним філіалом і виявіть, який філіал має найвищий.

In [8]:
prubytok_filial = df.groupby('Branch')['Total'].agg('sum').sort_values(ascending = False)
prubytok_filial 

Branch
C    110568.7065
A    106200.3705
B    106197.6720
Name: Total, dtype: float64

Найприбутковіший філіал С

2. В якому місті знайходиться філіал з найвищим прибутком? Може в тому місці нам розмітисти ще один магазин.  
Знайдіть відповідь за допомогою функціоналу Pandas.

In [13]:
prubytok_filial_city = df.groupby(['Branch','City'])['Total'].agg('sum').sort_values(ascending = False)
prubytok_filial_city

Branch  City     
C       Naypyitaw    110568.7065
A       Yangon       106200.3705
B       Mandalay     106197.6720
Name: Total, dtype: float64

Найприбутковіший філіал С в місті Naypyitaw

3.1. Створіть зводну таблицю, яка покаже, скільки покупок (інвойсів) було зроблено в кожній з філій (`Branch`) за різними категоріями товарів. Запишіть таблицю в змінну `invoices_by_category` і виведіть змінну на екран.
Ця таблиця допоможе проаналізувати, в якій філії купують найбільше товарів кожної з категорій.

In [19]:
invoices_by_catego = df.pivot_table(index = 'Branch', columns = ('Product line'), values = ('Invoice ID'), aggfunc ='count')
invoices_by_catego

Product line  Electronic accessories  Fashion accessories  Food and beverages  \
Branch                                                                          
A                                 60                   51                  58   
B                                 55                   62                  50   
C                                 55                   65                  66   

Product line  Health and beauty  Home and lifestyle  Sports and travel  
Branch                                                                  
A                            47                  65                 59  
B                            53                  50                 62  
C                            52                  45                 45

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1rueAdko6S3UxIHGtojetTxlES-EyM6Yb)

3.2. Викристовуючи змінну `invoices_by_category` дайте відповідь програмно (тобто значення треба не просто знайти очима, а вивести за допомогою коду), в якому філіалі магазину (`Branch`) найбільше інвойсів із покупкою товарів категорії "Електронні аксесуари" (`Electronic accessories`)?


In [21]:
invoices_by_catego['Electronic accessories'].idxmax()

'A'

4-6. **Творче завдання на розвиток аналітичного мислення**

Крок 1. Сформулюйте ТРИ питання (гіпотези) до наявних даних, які допомогли б вам зрозуміти, які користувачі що, де та коли найбільше/найменше купують, аби дати на основі цих гіпотез рекомендації бізнесу. Звісно питання мають бути не тими, на які ми вже відповіли в завданнях модулю.

Крок 2. Знайдіть відповіді на свої питання з допомогою функціоналу pandas.

Крок 3. Напишіть, як відповідь на це питання може бути використана для прийняття бізнес рішень.   
   
 Питання можуть бути будь-якої складності, але їх має бути 3. Кожне питання оцінюється як 1 завдання. Без виконання цього завдання ДЗ не приймається. Якщо є питання щодо виконання - пишіть у чат 🙌


1. Яка категорію товарів приносить найбільше прибутку серед жінок та чоловіків окремо?
2. Який тип оплати найбільше обирають покупці?
3. Визначити для кожного міста який тип клієнтів скільки робить покупки?

Перше завдання

In [22]:
category_gender = df.pivot_table(index = 'Gender', columns = ('Product line'), values = ('Total'), aggfunc ='sum')
category_gender

Product line  Electronic accessories  Fashion accessories  Food and beverages  \
Gender                                                                          
Female                    27102.0225            30437.400          33170.9175   
Male                      27235.5090            23868.495          22973.9265   

Product line  Health and beauty  Home and lifestyle  Sports and travel  
Gender                                                                  
Female               18560.9865          30036.8775         28574.7210  
Male                 30632.7525          23825.0355         26548.1055

In [30]:
top_category_female = category_gender.loc['Female'].idxmax()
print ('Найприбутковіша категорія товару для жінок: ',top_category_female, '\n')

top_category_male = category_gender.loc['Male'].idxmax()
print ('Найприбутковіша категорія товару для чоловіків: ',top_category_female)

Найприбутковіша категорія товару для жінок:  Food and beverages 

Найприбутковіша категорія товару для чоловіків:  Food and beverages


Відповідь: Як бачимо з отриманих результатів найбільше прибутку як для жінок, так і для чоловік є категорія Food and beverages. Саме ця категорія є ключовим 'драйвером' для магазинів, тому найбільше інвестицій потрібно залучати в дану категорію та наповнювати її новими позиціями, що дозволить збільшити чек даної категорії в рази.

Друге завдання

In [34]:
count_payment = df.groupby('Payment')['Invoice ID'].count()
persentg_payment = count_payment / sum(count_payment) * 100
persentg_payment

Payment
Cash           34.4
Credit card    31.1
Ewallet        34.5
Name: Invoice ID, dtype: float64

Результати показують, що різниця між типами оплат незначна і розприділена відсотково майже порівно. Якщо магазин захоче уникати готівки, то йому потрібно створювати різні акції, знижки,бонуси при оплаті онлайн або кредитною карткою, щоб залучити клієнтів обирати саме ці види оплат.

Третє завдання

In [42]:
totalby_city_member = df.pivot_table(index = 'City', columns = ('Customer type'), values = ('Invoice ID'), aggfunc ='count')
totalby_city_member

Customer type  Member  Normal
City                         
Mandalay          165     167
Naypyitaw         169     159
Yangon            167     173

З отриманих даних бачимо, що різниця між типом покупців незначна. Для того щоб стимулювати покупців ставати членами нашого магазину (member) потрібно створити рекламну кампанію, яка буде описмувати переваги членства. Запровадити акції для тих хто стане членом (member) нашого магазину, наприклад 15% знижка на перший чек ітд. Для типів клієнтів member запровадити акції, які будуть поширюватись лише на них і будуть значно привабливішими ніж не бути членом магазину, щоб стимулювати продажі та більше залучати клієнтів та ставати членами нашого магазину